src/librustc/middle/resolve_lifetime.rs

```diff
@@ -1475,12 +1475,61 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                         _ => None,
                     } {
                         debug!("id ={:?} span = {:?} name = {:?}", node_id, span, name);
-                        self.tcx.struct_span_lint_node(
+                        let mut err = self.tcx.struct_span_lint_node(
                             lint::builtin::UNUSED_LIFETIMES,
                             id,
                             span,
                             &format!("lifetime parameter `{}` never used", name)
-                        ).emit();
+                        );
+                        if let Some(parent_def_id) = self.tcx.parent(def_id) {
+                            if let Some(node_id) = self.tcx.hir.as_local_node_id(parent_def_id) {
+                                if let Some(Node::Item(hir_item)) = self.tcx.hir.find(node_id) {
+                                    match hir_item.node {
```

Comment:
I'd probably pull this `match` into a helper

Reply:
I know that similar code is elsewhere. I'm not against merging this, but we should probably sit down and look for duplication. I've written similar code in the past (although I can't find the right PR).

Also, I believe that [`get_generics`](https://github.com/rust-lang/rust/pull/51862/files#diff-2611048bb760f028c808ed4a14b07c00R667) might allow you to reduce a bit of boilerplate.

Reply:
[`binding_suggestion`](https://github.com/rust-lang/rust/pull/44782/files#diff-9d22f4d35c9bd98748f3e1a7cda1b3e7R836) is a similar helper method. Maybe we should just have a single file with all of these helpers to avoid having to dig for them every time...